### ISSUE TITLE
When moving from ->beginObject() ... etc to binson_encode results are different

### ISSUE THREAD

**Issue by DmSProject** [assaabloy-ppi/binson-php#13]
I've started to move away from the real low level way to create binson to using the binson_encode(associative-array) and have noticed some odd things.

Below is an example where I first create a payload in low level which works perfectly in the application, then for comparison, I create the same payload but from binson_encode from the same data. Jusst include binson.php and test it to see the output in the log.

I can't see if I'm doing anything different between the two ways.

//This:

```
<?php
require_once('bison.php');

$rand = random_bytes(16);
//(note that the same generated rand is used in both samples)
 
$buf2 = '';
$output = new BinsonWriter($buf2);
			
$output->objectBegin()
	->putName("rand")
	->putBytes($rand)
	->putName("contact")
	>objectBegin()
		->putName("A")
		->putString("pubkey")
		->putName("B")
		->putInteger(0)
		>putName("sigPub")
		>putBytes(hex2bin('9b20254303fe7cf17bdfdcd551104b399ac790b96d26ad60740ca9331c80a422'))
		->putName("encPub")
		->putBytes(hex2bin('debb2944ef2aeb00788c0b5f26a3c562cb814e68a1624f13338268bdb9f4d51f'))
		->putName("name")
		->arrayBegin()
			->putName("en")
			->putString("owner name")
		->arrayEnd()
		->putName("time")
		->putInteger(0)
	->objectEnd()
->objectEnd();

$retVal = array(bin2hex($output->toBytes()), bin2hex($rand));
$decoded = binson_decode($output->toBytes());
if (null !== binson_decode($output->toBytes())){
	error_log('old valid!');
}
error_log('old-as-decoded: '.print_r($decoded,true));
error_log('old bin obj: '.$output->toBytes());
error_log('old encoded: '.print_r($retVal,true));

//And this:

$arr = ["en"=>'owner name'];
$objStruct = [
	"rand"=>$rand,
	"contact" => [
		"A"=>"pubkey",
		"B"=>0,
		"sigPub"=>hex2bin('9b20254303fe7cf17bdfdcd551104b399ac790b96d26ad60740ca9331c80a422'),
		"encPub"=>hex2bin('debb2944ef2aeb00788c0b5f26a3c562cb814e68a1624f13338268bdb9f4d51f'),
		"name"=>$arr,
		"time"=>0
		]
	];
 
$binObj = binson_encode($objStruct);
$ret2 = array(bin2hex($binObj), bin2hex($rand));
if (null !== binson_decode($binObj)){
	error_log('new valid!');
}
error_log('new unencoded: '.print_r($objStruct,true));
error_log('new bin obj: '.$binObj);
error_log('new encoded: '.print_r($ret2,true));
?>

```
Returns different data.
Structures looks identical as I debug them, but output is different, for example looking at the resulting binary object you can see that the one built from an array lacks the rand value in the beginning.

**Comment by ppmag** [comment]
Thank you, Dan....  I'll check this ASAP.

**Comment by ppmag** [comment]
Two aspects of problems there:
1) low-level writer API don't sort key-val pairs at all - it's task of upper leyer - or API caller.
2) low-level writer API don't do any validation of context.
E.g. 
```php
->arrayBegin()
			->putName("en")
			->putString("owner name")
->arrayEnd()
```
will create array block, but object implies what specified later with `['en'=>'owner name']`

I will think how to make problem checks more robust.
Also I think in future no need to use low-level API.


So, please reorder fields in low-level part and fix `->arrayBegin()`  to `->objectBegin()`, etc,
for ['en'=>'owner name']  part
and it should work.

**Comment by DmSProject** [comment]
But hang on. 
It's the low level that does work as it is in the example here.
When doing the same thing using binson_encode from the array, it doesn't work.

**Comment by ppmag** [comment]
Yes, because:
1) binson_encode() reorder fields internally -> affect raw bytes

I will post fixed version soon.

**Comment by franslundberg** [comment]
Dan, the code

    $output->objectBegin()
	    ->putName("rand")
	    ->putBytes($rand)
	    ->putName("contact")
	    ->objectBegin()
            ....

Is *wrong*. "contact" must be before "rand" field. Otherwise you will not get valid Binson. Fields must be ordered in alphabetical order. Or more exact (from spec at binson.org): "Fields must be stored in order. The order must be the lexicographical order of the UTF-8 bytes of the name of the fields."

**Comment by franslundberg** [comment]
Alex, it would be good if Dan's example code would result in an *error* rather then it continuing to create bytes.

**Comment by DmSProject** [comment]
Frans.
Yes, I understand that the order is wrong.
What is interesting is that this snippet creates the payload for the QR/Deeplink, and it has worked even from this state.

But it will be ported to binson_encode which will ensure that the created order is correct.

**Comment by franslundberg** [comment]
In an old version of binson-java, invalid bytes (invalid order of fields) were accepted as Binson. It is fixed in the latest release of Binson. Key App likely use an older version of binson-java.

**Comment by DmSProject** [comment]
Ok. 
While waiting for a resolution to why binson_encode won't work for me I changed the order in the low-level, works fine whsen in the correct order, it did break when I did as Alex proposed and changed  ->arrayBegin() to ->objectBegin().

**Comment by ppmag** [comment]
Dan, I'll see.

**Comment by ppmag** [comment]
Frans, Dan, thank you for suggestion!

Proposed fixes related to current issue:
1) Low-level API: track last fieldname and exit with error, when out-of-order "->putName()"  requested.
2) Low-level API: exit with error, when "->putName()"  in ARRAY context requested.
2) Low-level API: exit with error, when illegal "->put*()" sequence  in OBJECT context requested.

3) High-level API:  by default, disable silent input array sorting inside binson_encode(). During encoding process, when out-of-order field found just exit with error (binson_encode() return null).  This will improve coding style by motivating library user to compose valid sorted arrays befor passing to binson_encode().  Also silent default reordering may cause 
(as we can see in this issue) more difficulties with search of the reason why output differs.
 For rare cases of misordered input data, it will be possible to force resorting by changing default behaviour with optional config parameter:

```php
$bytes = binson_encode($arr, ['serializer_sort_fields' => true]);
```

What you think?

**Comment by DmSProject** [comment]
I would be ok with needing to follow the correct order when setting up the array structures if it's needed in order to avoid odd and hard to trace errors.

But of course it would be nice if there was a reliable way to sort it automatically on creation.
I guess the optional flag would allow that, but then also introducing a potential error from the sorting.
Neither one is perfect, but I guess it's better to require a proper ordered array by default.

Frans, any thoughts?

**Comment by ppmag** [comment]
fixed in a019ed6
